### PR TITLE
feat: add meta commit task to defer commit meta data.

### PR DESCRIPTION
**Pull request by IRONICBo**
# Background

Current `write` method will update metadata frequently and cause a lot of metadata server overhead, we need to update the metadata info in meta task to reduce this cost.

In addition, this commit changes openfiles to move the maintenance of opencount to filehandle.

The separate readers and writers in filehandle are merged into handle, without distinguishing between separate read and write classes.

The meta commit operation is serialized into flush.

# Design 

1. Refactor storage filehandles to be shared.
2. Adjust truncate behavior and support implicit open/close for setattr_helper.
3. Add metacommit task for storage.

### CODE REVIEW

**Comment by IRONICBo** [comment]
Check tests:
```rs
test storage::storage_manager::test::concurrency::test_concurrency_aligned ... ok
test storage::storage_manager::test::concurrency::test_concurrency_unaligned ... ok
test new_storage::storage_manager::tests::storage_real_workload_test ... ok

test result: ok. 189 passed; 0 failed; 3 ignored; 0 measured; 0 filtered out; finished in 31.01s

   Doc-tests datenlord

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

**Comment by IRONICBo** [comment]
Version changed:

1M
```bash
daten-server:~/data$ dd if=/dev/zero bs=1M count=20 | tr '\0' '1' > datenlord_cache/new_storage_20m_test_version10.txt
20+0 records in
20+0 records out
20971520 bytes (21 MB, 20 MiB) copied, 10.5831 s, 2.0 MB/s
```
new storage with meta commit, version change is 12(10 period write and 1 flush/close request):
```bash
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":93272,"raft_term":2},"kvs":[{"key":"STI0MDAwMw==","create_revision":93261,"mod_revision":93272,"version":12,"value":"eyJOb2RlIjp7InBhcmVudCI6MSwibmFtZSI6Im5ld19zdG9yYWdlXzIwbV90ZXN0X3ZlcnNpb24xMC50eHQiLCJhdHRyIjp7ImlubyI6MjQwMDAzLCJzaXplIjoyMDk3MTUyMCwiYmxvY2tzIjowLCJhdGltZSI6eyJzZWNzX3NpbmNlX2Vwb2NoIjoxNzI1NDIyMzAyLCJuYW5vc19zaW5jZV9lcG9jaCI6MjM0MzgzNzA5fSwibXRpbWUiOnsic2Vjc19zaW5jZV9lcG9jaCI6MTcyNTQyMjMxMiwibmFub3Nfc2luY2VfZXBvY2giOjgxODQxNjk4MX0sImN0aW1lIjp7InNlY3Nfc2luY2VfZXBvY2giOjE3MjU0MjIzMDIsIm5hbm9zX3NpbmNlX2Vwb2NoIjoyMzQzODM3MDl9LCJraW5kIjoiUmVnIiwicGVybSI6NDM2LCJubGluayI6MSwidWlkIjoxMDIyLCJnaWQiOjEwMjQsInJkZXYiOjB9LCJkYXRhIjoiRmlsZSIsImxvb2t1cF9jb3VudCI6MSwiZGVmZXJyZWRfZGVsZXRpb24iOmZhbHNlfX0="}],"count":1}
```

new storage without meta commit, version change is 5120:
```bash
{"header":{"cluster_id":14841639068965178418,"member_id":10276657743932975437,"revision":93168,"raft_term":2},"kvs":[{"key":"STExMDAwMg==","create_revision":88049,"mod_revision":93168,"version":5120,"value":"eyJOb2RlIjp7InBhcmVudCI6MSwibmFtZSI6Im5ld19zdG9yYWdlXzIwbV90ZXN0X3ZlcnNpb25fb3JpZ2luLnR4dCIsImF0dHIiOnsiaW5vIjoxMTAwMDIsInNpemUiOjIwOTcxNTIwLCJibG9ja3MiOjAsImF0aW1lIjp7InNlY3Nfc2luY2VfZXBvY2giOjE3MjU0MTU1MTAsIm5hbm9zX3NpbmNlX2Vwb2NoIjo0MTU4Njg5Njh9LCJtdGltZSI6eyJzZWNzX3NpbmNlX2Vwb2NoIjoxNzI1NDE1NTE4LCJuYW5vc19zaW5jZV9lcG9jaCI6NjUxMDAzNDczfSwiY3RpbWUiOnsic2Vjc19zaW5jZV9lcG9jaCI6MTcyNTQxNTUxMCwibmFub3Nfc2luY2VfZXBvY2giOjQxNTg2ODk2OH0sImtpbmQiOiJSZWciLCJwZXJtIjo0MzYsIm5saW5rIjoxLCJ1aWQiOjEwMjIsImdpZCI6MTAyNCwicmRldiI6MH0sImRhdGEiOiJGaWxlIiwibG9va3VwX2NvdW50IjoxLCJkZWZlcnJlZF9kZWxldGlvbiI6ZmFsc2V9fQ=="}],"count":1}
```

File: src/new_storage/handle/writer.rs [hunk 1]
```diff
@@ -112,6 +139,24 @@ async fn write_back_block(task: Arc<WriteTask>) -> StorageResult<()> {
         }
     }
 
+    // Send the meta commit task.
+    let meta_task = Arc::new(MetaCommitTask { ino: task.ino });
+    match task
+        .meta_task_sender
+        .send(MetaTask::Pending(meta_task))
```
Comment (IRONICBo) [inline src/new_storage/handle/writer.rs:146]: Write channel sequence.

**Comment by IRONICBo** [comment]
> FileHandle.access() can be remove

OK, file level version PR will solve the block mismatch issue.

**Comment by IRONICBo** [comment]
truncate 20M:
version change: 10

write:
version change: 556

File: src/async_fuse/memfs/s3_metadata.rs [hunk 2]
```diff
@@ -176,43 +167,22 @@ impl MetaData for S3MetaData {
...
     }
 
     #[instrument(skip(self), err, ret)]
-    async fn open(&self, context: ReqContext, ino: u64, flags: u32) -> DatenLordResult<u64> {
+    async fn open_remote(
+        &self,
+        context: ReqContext,
+        ino: u64,
+        flags: u32,
+    ) -> DatenLordResult<(u64, FileAttr)> {
```
Comment (frostyplanet) [inline src/async_fuse/memfs/s3_metadata.rs:185]: I don't think it's nessecary to name it 'remote'

File: src/async_fuse/memfs/s3_metadata.rs [hunk 3]
```diff
@@ -176,43 +167,22 @@ impl MetaData for S3MetaData {
...
-            res?;
-        }
-        Ok((file_size, mtime))
+    async fn open_local(
+        &self,
+        _context: ReqContext,
+        _ino: u64,
+        _flags: u32,
+    ) -> DatenLordResult<u64> {
+        return Ok(self.allocate_fd());
```
Comment (frostyplanet) [inline src/async_fuse/memfs/s3_metadata.rs:176]: It seems the function is not called

File: src/new_storage/storage_manager/handle.rs [hunk 4]
```diff
@@ -414,81 +340,189 @@ impl FileHandleInner {
...
+                            let slf_clone = Arc::clone(&self);
+                            let res = slf_clone.write_blocks(&tasks, Arc::clone(&metadata_client)).await;
+                            tasks.clear();
+                            let res = result.take().or(res);
+                            if let Err(Some(e)) = tx.send(res) {
+                                error!("Failed to send storage error back to `Writer`, the error is {e}.");
+                            }
+                        }
+                        Task::Finish(tx) => {
+                            info!("Finish the write back task. ino: {:?}", self.ino);
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:458]: Finish & Flush code branch mostly the same, should be merged

**Review by frostyplanet — CHANGES_REQUESTED** [review]

File: src/async_fuse/memfs/mod.rs [hunk 5]
```diff
@@ -229,16 +227,36 @@ impl<M: MetaData + Send + Sync + 'static> FileSystem for MemFs<M> {
...
-            Ok((ttl, fuse_attr)) => {
+
+        // Get from local storage first
+        if let Ok(file_attr) = self.storage.getattr(ino).await {
+            debug!(
+                "getattr() successfully got the attr={:?} of ino={}",
+                file_attr, ino,
+            );
+            // TODO: Mock response ttl.
+            let ttl = Duration::new(3600, 0);
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/async_fuse/memfs/mod.rs:243]: Ok, done.

File: src/async_fuse/memfs/s3_metadata.rs (hunk 2 above)
Comment (IRONICBo) [reply src/async_fuse/memfs/s3_metadata.rs:185]: Compared to `open_local` method, this method will get remote attr from kvengine.

File: src/async_fuse/memfs/s3_metadata.rs (hunk 3 above)
Comment (IRONICBo) [reply src/async_fuse/memfs/s3_metadata.rs:176]: Ok, this method is removed and replaced by `open`.

File: src/new_storage/storage_manager/handle.rs [hunk 6]
```diff
@@ -551,13 +561,67 @@ impl FileHandle {
...
+        let _guard = self.open_close_lock.lock().await;
+        println!("try to get open filehandle lock ok {:?}", self.fh);
         self.inner.open();
     }
 
+    /// Get and try to increase the open count of the file handle.
+    pub async fn get_and_open(&self) -> bool {
+        let _guard = self.open_close_lock.lock().await;
+        println!("try to get open filehandle lock ok {:?}", self.fh);
+        if self.inner.open_cnt() > 0 {
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:574]: no need to open_count > 0,  just increase open_count will do

**Review by frostyplanet — CHANGES_REQUESTED** [review]

File: src/new_storage/storage_manager/handle.rs [hunk 7]
```diff
@@ -601,68 +666,135 @@ impl Handles {
...
+    fn get_shard(&self, fh: u64) -> &Arc<tokio::sync::RwLock<HashMap<u64, FileHandle>>> {
         let idx = Self::hash(fh);
         self.handles
             .get(idx)
             .unwrap_or_else(|| unreachable!("The array is ensured to be long enough."))
     }
 
     /// Adds a file handle to the collection.
-    pub fn add_handle(&self, fh: FileHandle) {
+    pub async fn add_handle(&self, fh: FileHandle) {
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:677]: add_handle has no use

File: src/new_storage/storage_manager/handle.rs [hunk 8]
```diff
@@ -601,68 +666,135 @@ impl Handles {
...
-        let mut shard_lock = shard.write();
-        shard_lock.push(fh);
+        let mut shard_lock = shard.write().await;
+        shard_lock.insert(fh.fh(), fh);
     }
 
     /// Removes a file handle from the collection.'
     #[must_use]
-    pub fn remove_handle(&self, fh: u64) -> Option<FileHandle> {
+    pub async fn remove_handle(&self, fh: u64) -> Option<FileHandle> {
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:685]: remove_handle is not called

File: src/new_storage/storage_manager/handle.rs [hunk 9]
```diff
@@ -98,7 +90,7 @@ enum Task {
 
 /// The `WriteTask` struct represents a write task
 #[derive(Debug)]
-struct WriteTask {
+pub struct WriteTask {
     /// The cache manager.
     cache: Arc<Mutex<MemoryCache<CacheKey, LruPolicy<CacheKey>>>>,
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:95]: "cache, block, ino" is duplicated with FileHandleInner,  should be removed

File: src/async_fuse/memfs/mod.rs [hunk 10]
```diff
@@ -12,7 +12,7 @@ pub mod direntry;
 mod metadata;
 mod node;
 /// Opened files
-mod open_file;
+// pub mod open_file;
```
Comment (frostyplanet) [inline src/async_fuse/memfs/mod.rs:15]: Delete open_file.rs since it's function is replaced, do not leave obsoleted code

**Review by frostyplanet — CHANGES_REQUESTED** [review]

File: src/new_storage/storage_manager/handle.rs [hunk 11]
```diff
@@ -551,13 +561,67 @@ impl FileHandle {
...
     /// Increase the open count of the file handle.
-    pub fn open(&self) {
+    pub async fn open(&self) {
+        let _guard = self.open_close_lock.lock().await;
+        println!("try to get open filehandle lock ok {:?}", self.fh);
         self.inner.open();
     }
 
+    /// Get and try to increase the open count of the file handle.
+    pub async fn get_and_open(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs]: Ok, removed.

File: src/new_storage/storage_manager/handle.rs [hunk 12]
```diff
@@ -551,13 +561,67 @@ impl FileHandle {
...
+            return true;
+        }
+
+        false
+    }
+
+    /// Get the open count of the file handle.
+    /// Be careful, do not split the `open_cnt` with open and close operation.
+    #[must_use]
+    pub async fn open_cnt(&self) -> u32 {
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs]: Ok.

File: src/async_fuse/memfs/mod.rs [hunk 13]
```diff
@@ -229,16 +227,36 @@ impl<M: MetaData + Send + Sync + 'static> FileSystem for MemFs<M> {
...
+            let fuse_attr = fs_util::convert_to_fuse_attr(file_attr);
+            debug!(
+                "getattr() successfully got the attr={:?} of ino={}",
+                fuse_attr, ino
+            );
+            return reply.attr(ttl, fuse_attr).await;
+        }
+
+        // Get from remote metadata server
+        match self.metadata.get_remote_attr(ino).await {
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/async_fuse/memfs/mod.rs:253]: @frostyplanet  But the `storage.getattr()` will fetch metadata from the cache, and `metadata.get_remote_attr(ino)` will fetch metadata from engine; they are in different behavior.

File: src/new_storage/storage_manager/handle.rs (hunk 7 above)
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs:688]: Ok, `add_handle` is used in ut, changed to `create_or_open_handle`.

File: src/new_storage/storage_manager/handle.rs (hunk 9 above)
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs:97]: Ok, delete `ino` and change the function to FIleHandleInner.

File: src/new_storage/storage_manager/handle.rs [hunk 14]
```diff
@@ -414,80 +338,272 @@ impl FileHandleInner {
...
+            if let Err(e) = metadata_client
+                .write_remote_size_helper(ino, file_size)
+                .await
+            {
+                error!("Failed to commit meta data, the error is {e}.");
+            }
+
+            let mut attr_write = self.attr.write().await;
+            if let Some(dirty_filesize) = attr_write.dirty_filesize {
+                if dirty_filesize <= file_size {
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:465]: I think should be `if dirty_filesize == file_size`. 
Set self.attr directly is not bulletproof,  should use latest attr query from  write_remote_size_helper after update suc

File: src/new_storage/storage_manager/handle.rs (hunk 4 above)
Comment (frostyplanet) [reply src/new_storage/storage_manager/handle.rs:541]: Remove Task::Finish, no used any more

File: src/new_storage/storage_manager/handle.rs [hunk 15]
```diff
@@ -414,80 +338,272 @@ impl FileHandleInner {
...
+                            }
+
+                            // Check last write task size with mem attr
+                            return;
+                        }
+                    }
+                }
+                _ = interval.tick() => {
+                    let slf_clone = Arc::clone(&self);
+                    slf_clone.write_blocks(&tasks, Arc::clone(&metadata_client)).await;
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:557]: If len(tasks)==0, should not call write_blocks

File: src/new_storage/storage_manager/handle.rs [hunk 16]
```diff
@@ -414,80 +338,272 @@ impl FileHandleInner {
...
+                    // Update current file size.
+                    attr_write.attr.size = file_size;
+                }
+            }
+
+            return result;
+        }
+
+        let mut attr_write = self.attr.write().await;
+        if let Some(dirty_filesize) = attr_write.dirty_filesize {
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/handle.rs:477]: Delete this section,  should not use self.attr to update meta

**Review by frostyplanet — CHANGES_REQUESTED** [review]

File: src/new_storage/storage_manager/handle.rs (hunk 14 above)
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs]: OK, I have updated the filesize with remote size.

File: src/new_storage/storage_manager/handle.rs (hunk 15 above)
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs:530]: OK, I find that `write_blocks` will quickly return if the task is none, so we can keep it.

File: src/new_storage/storage_manager/handle.rs (hunk 16 above)
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs]: Ok, removed.

File: src/new_storage/storage_manager/handle.rs [hunk 17]
```diff
@@ -414,80 +336,247 @@ impl FileHandleInner {
...
-            > 0
+            > 1
         {
-            return Ok(());
+            debug!("The file handle is still open by other processes.");
+            return Ok(false);
         }
 
-        let (tx, rx) = oneshot::channel();
+        debug!("The file handle is closed, remove the file handle.");
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs:359]: OK, move to `try_close`.

File: src/new_storage/storage_manager/handle.rs [hunk 18]
```diff
@@ -414,80 +336,247 @@ impl FileHandleInner {
...
+    #[must_use]
+    pub fn get_open_count(&self) -> u32 {
+        self.open_cnt.load(std::sync::atomic::Ordering::SeqCst)
+    }
+
     /// Closes the writer associated with the file handle.
-    pub async fn close(&self) -> StorageResult<()> {
+    /// If the open count is greater than 0, it will return false and do not remove this handle.
+    /// Otherwise, it will return true and remove this handle.
+    pub async fn close(&self) -> StorageResult<bool> {
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs:348]: Done.

File: src/new_storage/storage_manager/handle.rs [hunk 19]
```diff
@@ -414,80 +336,247 @@ impl FileHandleInner {
...
-            });
+        // Check dirty count to determine whether the file handle is clean.
+        self.flush().await?;
 
-        {
-            let keys = self.access_keys.lock();
-            for key in keys.iter() {
-                self.cache.lock().remove(key);
+        // Will be closed by the send method, or the sender is dropped.
+        self.close_sender.send(()).await.unwrap_or_else(|_| {
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs:365]: Ok, that make sense.

File: src/new_storage/storage_manager/handle.rs [hunk 20]
```diff
@@ -601,68 +710,120 @@ impl Handles {
...
         let shard = self.get_shard(fh);
-        let shard_lock = shard.read();
-        let fh = shard_lock.iter().find(|h| h.fh() == fh)?;
-        Some(fh.clone())
+        let mut shard_lock = shard.write().await;
+        let filehandle = shard_lock.get(&fh).ok_or_else(|| {
+            StorageError::Internal(anyhow::anyhow!("Cannot close a file that is not open."))
+        })?;
+
+        let need_remove_flag = filehandle.close().await?;
```
[the remark this answers is not in the document]
Comment (IRONICBo) [reply src/new_storage/storage_manager/handle.rs:777]: Done.

**Review by frostyplanet — APPROVED** [review]

**Pull request merged**